cpp/src/common/logger.cpp

```diff
@@ -48,7 +50,10 @@ Logger& Logger::get() {
   return logger;
 }
 
-Logger::Logger() : logger{spdlog::stdout_color_mt("cuml")}, currPattern() {
+Logger::Logger()
+  : sink{std::make_shared<spdlog::sinks::callback_sink_mt>()},
```

Comment:
I looked at this before, and the implementation got ugly, because the formatter was declared private and we would need it in the overriding `log` function. It would also mean that this implementation would follow some unusual patterns for a custom sink, since `ansicolor_sink` does not follow the typical pattern for custom sinks.

*But* as I was writing up this response, I thought of a fairly straightforward workaround for the worst part of the ugliness. I'm going to give it one more go right now, and I'll either push the result or post here if it falls apart again.

Reply:
All right, I've now convinced myself that the workaround will not work.

- `stdout_color_mt` is the factory for `stdout_color_sink_mt`.
- `stdout_color_sink_mt` is an alias for `ansicolor_sink<details::console_mutex>`
- We need access to `formatter_` to correctly format the output in the overriding `log` function of `CallbackSink`
- `formatter_` is a unique_pointer which is declared private and has no protected accessor
- As a workaround, we might consider defining a separate formatter member in the derived class and overriding `set_pattern` to update the pattern for this separate formatter and then  calling the parent set_pattern as well, but...
  - `set_pattern` is declared final
- As another workaround, we might consider defining the separate formatter member in the derived class and then using that wherever the parent uses `_formatter`, but...
  - We would have to fully copy the `log` logic (and a bunch of other stuff) from the parent, at which point the inheritance doesn't really buy us anything

The other thing I considered previously was that we could _wrap_ an instance of `ansicolor_sink` rather than inheriting from it. That seems entirely viable to me, but it's also a little ugly (leading to my earlier "juice may not be worth the squeeze" comment). I'm actually less perturbed by that approach than I was when I was originally thinking about this, but it may still not be worth it if we're not using any `ansicolor_sink`-specific features? I don't know; I'm open to others' opinions on that.